src/package.rs

```diff
@@ -373,52 +424,62 @@ impl Package {
-        // archive.
-        let tarfile = self.get_output_path(name, output_directory);
-        let file = open_tarfile(&tarfile)?;
-
-        // TODO: Consider using async compression, async tar.
-        // It's not the *worst* thing in the world for a packaging tool to block
-        // here, but it would help the other async threads remain responsive if
-        // we avoided blocking.
-        let gzw = flate2::write::GzEncoder::new(file, flate2::Compression::fast());
-        let mut archive = Builder::new(gzw);
-        archive.mode(tar::HeaderMode::Deterministic);
-
-        // The first file in the archive must always be a JSON file
-        // which identifies the format of the rest of the archive.
-        //
-        // See the OMICRON1(5) man page for more detail.
-        let mut root_json = tokio::fs::File::from_std(tempfile::tempfile()?);
-        let contents = r#"{"v":"1","t":"layer"}"#;
-        root_json.write_all(contents.as_bytes()).await?;
-        root_json.seek(std::io::SeekFrom::Start(0)).await?;
-        archive.append_file("oxide.json", &mut root_json.into_std().await)?;
-
-        // Add mapped paths.
-        self.add_paths(progress, &mut archive).await?;
-
-        // Attempt to add the rust binary, if one was built.
-        progress.set_message("adding rust binaries");
-        if let Some(rust_pkg) = &self.rust {
-            let dst = Path::new("/opt/oxide").join(&self.service_name).join("bin");
-            add_directory_and_parents(&mut archive, &dst)?;
-            let dst = archive_path(&dst)?;
-            rust_pkg.add_binaries_to_archive(&mut archive, &dst)?;
-            progress.increment(1);
+        let mut archive = new_zone_archive_builder(name, output_directory).await?;
+
+        match &self.source {
+            PackageSource::Local { paths, .. } => {
+                // Add mapped paths.
+                self.add_paths(progress, &mut archive, &paths).await?;
+
+                // Attempt to add the rust binary, if one was built.
+                self.add_rust(progress, &mut archive).await?;
+
+                // Add (and possibly download) blobs
+                let blob_dst = Path::new("/opt/oxide").join(&self.service_name).join(BLOB);
+                self.add_blobs(
+                    progress,
+                    &mut archive,
+                    output_directory,
+                    &archive_path(&blob_dst)?,
+                )
+                .await?;
+            }
+            PackageSource::Composite { packages } => {
+                // For each of the component packages, open the tarfile, and add
+                // it to our top-level archive.
+                let tmp = tempfile::tempdir()?;
+                for component_package in packages {
+                    let component_path = output_directory.join(component_package);
+                    let gzr = flate2::read::GzDecoder::new(open_tarfile(component_path)?);
```

Comment:
Are the members of a composite package expected/required to be zone packages and not tarball packages? (I think this will fail if `component_package` is a tarball package (not gzipped), right?) Should we check that explicitly to give an earlier/better error message?

Reply:
Yeah, I went ahead and added an early-check here. Done.